enhancements/windows-containers/health-management.md

```diff
@@ -0,0 +1,337 @@
+  take the worker ignition file as an input, and will not involve the services ConfigMap.
+* WMCO will invoke the `controller` command to run WICD as a Windows service on the instance.
+* WMCO will invoke the `cleanup` command of WICD when removing an instance from the cluster. WMCO will no longer
+  delete the Node object.
+
+The Windows Node creation workflow for the end user will not be changed.
+
+### User Stories
+
+User stories can be found within the health management epic, [WINC-657](https://issues.redhat.com/browse/WINC-657).
+
+### API Extensions
+
+N/A
+
+### Risks and Mitigations
+
+* It is possible that issues with Windows instances will occur which WICD is unable to resolve, causing the Node to
+  enter a `NotReady` state. For instances created through the machine-api, the user can set up MachineHealthChecks,
+  which will correct the state by recreating the Machine. For BYOH instances, user intervention is required at this
+  point. Events and alerts will ensure that a cluster administrator is made aware of the issue.
+* There is a possibility that a cluster admin could mistakenly remove the services ConfigMap for an older WMCO version,
+  before Node upgrades are completed. This would result in WICD not knowing what services it has configured on the
+  instance. In order to prevent this, WICD can set the Windows service description to "openshift-managed", so that it
+  can search for services which were installed by WICD. As it is potentially possible for a user to remove this tag
+  from the services, when the expected service ConfigMap is missing, both the tag and the latest service ConfigMap
+  should be used when deconfiguring the node. Using a combination of these two will allow for all the installed
+  services to be removed, in most cases. Additionally, in order to facilitate upgrades from a version without
+  WICD, this tag should be added to all existing services in 4.10, one major version ahead of WICD's target release.
+
+## Design Details
+
+### Services ConfigMap
+
+WMCO will create and maintain an immutable ConfigMap with the naming scheme:
+`services-<MajorVersion>-<MinorVersion>-<PatchVersion>-<Commit>`. For example, if WMCO was built from commit a7b5
+of WMCO version 5.0.0, the created ConfigMap would be named `services-5-0-0-a7b5`.
+
+The purpose of this ConfigMap is to provide WICD with the specifications required to configure a Node. The ConfigMap
+will be watched via the existing ConfigMap controller. WMCO is the owner of the ConfigMap and the source of the data
+present within it. If the ConfigMap is deleted, WMCO will re-create it.
+
+If an entity attempts to create a services ConfigMap with incorrect values, WMCO will delete it.
+This will resolve a scenerio in which the ConfigMap controller is busy configuring Windows instances, and a user
+deletes and recreates the service ConfigMap with incorrect values, before the controller has a chance to re-create
+the ConfigMap.
+
```

Comment:
Eventual consistency, when it is no longer busy it will reconcile to the correct state.

Reply:
But what if WICD (on a Node  which is not being processed) which also looks at this ConfigMap reacts to the incorrect changes?

Reply:
The same result, the instance could be temporarily set up with an incorrect configuration, but it will be reverted when the ConfigMap is changed, and there will be eventual consistency.

There shouldn't be an expectation of WMCO to prevent these scenarios.

What can be done, is we can reduce the impact by allowing parallel reconciliations within the ConfigMap controller. This will reduce the reaction time of WMCO, allowing for the BYOH ConfigMap events to be processed at the same time as service events. This doesn't reduce the chance of this happening to zero, and there is nothing stopping a user from setting up an infinite loop of delete and create API calls to continuously attempt to set the ConfigMap how they want. RBAC is the proper way to deal with this, users should not be granted permissions to change the ConfigMap. In the end, if the cluster Admin wishes to mess with OpenShift's functionality with this level of disruption, that's on them.

Would allowing parallel reconciliations within the ConfigMap controller be enough of a mitigation in your opinion?

Reply:
>What can be done, is we can reduce the impact by allowing parallel reconciliations within the ConfigMap controller. 

Why should it be parallel reconciliations? You can have a semaphore that controls the number of reconciliations. Or have a separate thread outside of the controller that watches for deletion events? Either way this needs to be called out as a risk with a mitigation.

Reply:
Are there other examples of controllers deleting user-provided configuration instead of ignoring it and surfacing the error?

Reply:
CVO 😃 The services `ConfigMap` is similar to a CVO managed manifest. We expect it to look a certain way and will stomp on any user introduced changes.